### ISSUE TITLE
Feature request - native tile map support

### ISSUE THREAD

**Issue by iliatimofeev** [vega/vega#1212]
Since vega3 and vegalight2 supports complex charts with linked "subplots" it will be great to have geoshape with a tile map (OSM, Stamen) on background as "subplot" of the diagram.

Tile maps have fixed projection for each tile provider (in most cases it is Mercator). Thus tile map settings could be a part of projection properties.

Or tile map server could be implemented as data source so tile will be a mark type. 

Sample d3 implementation by Mike Bostock http://bl.ocks.org/mbostock/5342063 in my understanding is compatible with current implementation of zooming.

Another implementation http://bl.ocks.org/mbostock/5914438 is faster but may need some adjustment of zooming mechanism.

**Comment by jheer** [comment]
The tricky part here is that a Vega view renders into a single SVG or Canvas element. So, under the current scheme having embedded tile maps as part of the native spec would require internal mechanisms for copying the tile image data into the Vega scenegraph. This might well be workable (ignoring performance considerations for a moment), but would require significant development that is not currently on our roadmap.

**Comment by nyurik** [comment]
@iliatimofeev you may find https://github.com/nyurik/leaflet-vega a useful alternative.

**Comment by jheer** [comment]
This is fantastic! Thank you sharing @iliatimofeev. It is always such a treat for a tool builder to be surprised about what is possible with the tool...

I think it makes sense to encapsulate as much of the processing as possible (certainly all the transforms under the `tile_list` data source) within a new `tile` transform. Perhaps we can add a new vega-tile submodule to the Vega org to develop this.

I haven't taken a deeper look yet, but I think the "flashing" of tiles may be due to latency in image tile loading. My understanding is that tile managers like Leaflet are smart about adding / removing tiles, sometimes including interpolated transitions, whereas Vega will just immediately remove images no longer in the scenegraph, even if the new images haven't finished loading yet. Fixes here would probably involve some modifications to the tile loading strategy (e.g., keep old tiles arounds but draw newer tiles on top of them) or supporting tile pre-fetching to reduce load times. This all seems pretty reasonable.

I also wonder about how standardized tile URL reference systems are across tile providers. How many different tile providers can/should this support?

Were you interested in further developing this, or were you hoping to "hand off" to other folks involved in Vega development? Thanks again!

**Comment by nyurik** [comment]
Awesome!  Can this be moved to a github file/repo/wiki page?

`    {"name": "dx", "update": "round((floor(dii)-dii)*tileSize)"},`
should be
`    {"name": "dx", "update": "round((di-dii)*tileSize)"},`

**Comment by iliatimofeev** [comment]
@jheer thank you for so warm response. I made some comments on your questions and I hope it could help to move ahead with this feature that I so dream about. 😃 

>I also wonder about how standardized tile URL reference systems are across tile providers.

According [OGC WMTS Simple Profile](http://docs.opengeospatial.org/is/13-082r2/13-082r2.html#17)  implementations of simple RESTful clients are provided with tile-URL templates as inputs, with all but {TileMatrix}, {TileCol} and {TileRow} resolved. Clients assume a Web Mercator TileMatrixSet (or a World CRS84 TileMatrixSet) which is defined in profile. In my example `zoom=TileMatrix, i=TileCol, j=TileRow` as provided by *Stamen Design*. In [Annex D.1](http://docs.opengeospatial.org/is/13-082r2/13-082r2.html#33) you can find examples of templates like:
```
http://a.tile.openstreetmap.org/{TileMatrix}/{TileCol}/{TileRow}.png
http://mt1.google.com/vt/hl=en&z={TileMatrix}&y={TileRow}&x={TileCol}
http://www.ogc.uab.cat/SITiled/GeoEye_Haiti_M/GeoEye/WorldSphericalMercator/{TileMatrix}/{TileCol}/{TileRow}.jpg
http://maptile.maps.svc.ovi.com/maptiler/maptile/newest/normal.day/15/{TileCol}/{TileRow}/256/png8
```
[Leafletjs](http://leafletjs.com/reference-1.3.0.html#tilelayer) use extended template `'http://{s}.somedomain.com/blabla/{z}/{x}/{y}{r}.png'` where {s} means one of the available subdomains (used sequentially to help with browser parallel requests per domain limitation; subdomain values are specified in options; a, b or c by default, can be omitted), {z} — zoom level, {x} and {y} — tile coordinates. {r} can be used to add `@2x` to the URL to load retina tiles.
In fact it supports something like  `"https://cartodb-basemaps-{switch:a,b,c}.global.ssl.fastly.net/light_all/${z}/${x}/${y}@2x.png"`

> How many different tile providers can/should this support?

Technically this template for the url format is enough to work with most of services. For some commercial services like *google maps* we'll need to support some authentification mechanism. You can see popular providers comparation at  [bbbike.org](https://mc.bbbike.org/mc/) to understand scale of diversity on this market. 😄  
To improve user experience we can predefine some helpful services providers for data visualization. Personally I like [stamen-light](http://maps.stamen.com/toner-lite) which was designed for that usage and supplied under [CC BY 3.0 licence](https://creativecommons.org/licenses/by/3.0/) .   [Carto](https://carto.com/location-data-services/basemaps/) and [mapbox light & dark](https://www.mapbox.com/maps/light-dark/) are good to but licence do not allow private use. 

>Were you interested in further developing this, or were you hoping to "hand off" to other folks involved in Vega development?

As Altair user I will be honored to contribute to Vega, unfortunately I am not really javascript developer. I could try under some guidance especially on how to deploy development environment. 

> I think the "flashing" of tiles may be due to latency in image tile loading.

As I see there are two types of flashing.
First when you move map out of current tiles bounding box so it needs extra column or row, in that case `di` or `dj` will change and new images will be assigned to all data point. I think more careful data point management (just add new and remove unused) will help. 
Second when you scale map and it has to switch to next TileMatrix in that case problem as you describe it.  It could be solved by removing "old" data points only after successful loading replacement images.

In my opinion for data visualization both is not critical scenarios.  Static map are more common in charts, but may be it because of current tools restriction.

**Comment by nyurik** [comment]
@iliatimofeev I think there is little value in a built-in support of a "tile" concept in Vega, because you would very rarely use it without the rest of the code you wrote above (again, very impressive!).  Either the user would need "the whole thing" (e.g. Leaflet, Open Layers, or Mapbox -like functionality) or none of it.  But if the user needs the whole thing, they are much better off using the whole library specifically made for tiles, with a huge community support, performance and GPU optimizations, ongoing maintenance by a large team of devs, etc.

By now, every map end-user expects similar functionality from a map:  smooth scrolling and zooming using drags and wheel scrolls, plus possibly a number of other, lesser known features for accessibility.  What would be the point of re-implementing it in raw Vega from scratch, except for the demo purposes?  I might be mistaken of course, and there is a good usecase :)

**Comment by jheer** [comment]
I agree with @nyurik that Vega should not try to provide support equivalent to Leaflet, etc. However, I also think relatively simple native tile support could be quite useful, even if not as performant. In particular, I expect Vega-Lite and Altair users might greatly appreciate that support.

**Comment by domoritz** [comment]
> I expect Vega-Lite and Altair users might greatly appreciate that support.

Yes! I think in most cases users would want a map that doesn't move or zoom. However, having support for zooming if desired would be pretty awesome. Here are a few map examples in Vega-Lite: https://vega.github.io/vega-lite/examples/#geographical-maps. One concern I have is that tile maps usually only work with the Mercator projection and so we would need to prevent users from using tiles with other projections.

**Comment by nyurik** [comment]
If the desire is to have a background map, than I think it might be better to fetch a single image snapshot of a given area. E.g.  https://maps.wikimedia.org/img/osm-intl-i18n,7,37.8013,-122.3988,500x500.png?lang=ru

**Comment by iliatimofeev** [comment]
@nyurik I understand your skepticism, given the availability of a working solutions to display maps in your use cases. And I agree that zooming and positioning on map it is what end user expects it is really important. But if I have to choose between having a snapshot on background or nothing, i'll vote for snapshot.

I would like to describe my class of **use cases** and then we could see a better way to implement them.

#### I use jupyter notebook for data exploration and analysis on python. 

From the technical viewpoint in my case there is some stack above the Vega. Altair generates Vega-light specification, then Jupyter renderer pass it to vega-embed which calls Vega-light to transform it to Vega which renders using d3.js and here I get my pretty and interactive image in the output cell.

@jakevdp promoted Altair as simple wrapper over promising technology Vega\Vega-light with key feature as descriptive specification of interactions. Since Vega-light 2, it allows to produce a set of interactive co-depend charts, like a dashboard. It's really outstanding feature I don't know another tools could do it in Jupyter Notebook. 

#### Imagine dataset with time and spatial information.
Something like: taxi call log, cellular operators logs, bank cards transactions log, advisement clicks or feeds of twitter or instagram. All of it has time component and location data (or could be enriched with it simply)  but a lot of other valuable information that could be analyzed. Time and location are key predictors but also i'm interested in some visualization of what is going on with other factors on that time interval in that area. So i would like to see chart and map with shared selection and legends. 
#### Something like this picture but where you can select a point on the map to see where it on the chart and otherwise (cell operator log).
<img width="813" alt="point_map" src="https://user-images.githubusercontent.com/11561581/39359966-df91905a-4a24-11e8-9178-0dbf0d2fbc6e.png">
<img width="813" alt="points_chart" src="https://user-images.githubusercontent.com/11561581/39359915-ad976ff2-4a24-11e8-82e5-0522e80cdfa3.png">

Connecting this two views could give a lot of answers that is difficult to understand other ways: are they (blue and red) live together, work same place, friends who meet sometimes or just two persons who coincidentally stuck in one traffic jam. It is really valuable information.

#### Another example based on NYC taxi call data
![vega 2](https://user-images.githubusercontent.com/11561581/39360999-2d9ed74a-4a29-11e8-8941-c7c237782947.png)

[vl2 interactive version](https://iliatimofeev.github.io/nytaxi/dashboard.html)  on my github it would not react on time selection but it is other issue and i'll describe it on vega-light.

### Back to use cases 

There is a huge gap between chart tools and GIS tools they live in a different worlds and I see a great power in connecting them and real potential in Vega/Vega-light to do it. @domoritz already did really amazing work in that direction which allows to do complex things like NYC taxi calls with little efforts from my side. I think i spent more time on searching and preparing background map of NYC than on preparing this dashboard. 

1. Simple background map for spatial data where data items selection could affect on other charts it is huge step ahead and time saver for me. 
2. Scaling and moving it is second thing to ask. Map content is dependent on current scale and data point could be placed in the middle of nowhere on current scale and we need to change zoom to understand where it is.
3. Switching tile provider online could be useful sometimes because different tiles could have different content. Rich ones are ugly as background but helpful for analysis. 
4. Support of vector tiles like mapbox its a future of web GIS systems but for today is to far away to be a basic need.  


### Implementation approach

My first suggestion was to use a d3.js implementation of tile maps - yes d3.js has it and it is [80 lines of code](https://github.com/d3/d3-tile/blob/master/src/tile.js) but as @jheer  said it can't be used in dashboards because it use full canvas not a part as Vega need. I suppose that current Vega compatible implementations of tiles do the same. 

Well Altair by a concept is only a wrapper , Vega-light by a concept is only a transformation to Vega. So the only reasonable place to ask for tile map support is here where I am asking.

**Comment by mattijn** [comment]
This is insane. Really great! Next to WMTS there is also WMS, TMS and XYZ.

But since WMTS  is the only integer index tile protocol backed by a widely recognized standards body and also having the best performance in production environments it would be sufficient to support WMTS alone.

Even though there is some blinking I'm impressed by the performance already

**Comment by mattijn** [comment]
In my environment the question how to implement this (albeit in Altair) is coming back on a weekly basis. Really like to have this possibility.

**Comment by djouallah** [comment]
End user here,  I use both leaflet and Altair, ideally if crossfiltering works between the two libraries, it will be a game changer!!
